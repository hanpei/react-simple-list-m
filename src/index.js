import List from './list/List';
import HorizontalList from './list/HorizontalList';

export default List;
export { HorizontalList };
