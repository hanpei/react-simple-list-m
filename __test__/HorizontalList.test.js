import React from 'react';
import { shallow, mount } from 'enzyme';

import { HorizontalList as List } from '../src';

describe('HorizontalList', () => {
  const data = (() => {
    const arr = [];
    for (let i = 0; i < 30; i += 1) {
      arr.push(i.toString());
    }
    return arr;
  })();

  it('render', () => {
    const props = {
      items: data,
      itemWidth: 60,
      indexKey: (item) => Number(item),
      renderItem: (item) => <div>{item.toString()}</div>,
    };
    const nextProps = {
      items: data.concat([31, 32, 33, 34]),
      itemWidth: 60,
      indexKey: (item) => Number(item),
      renderItem: (item) => <div>{item.toString()}</div>,
    };
    const component = shallow(<List {...props} />);
    expect(component).toMatchSnapshot();
    component.setProps(nextProps);
    expect(component).toMatchSnapshot();
    component.unmount();
    expect(component).toMatchSnapshot();
  });

  it('render loading', () => {
    const fn = jest.fn();
    const props = {
      items: data,
      itemWidth: 60,
      indexKey: (item) => Number(item),
      renderItem: (item) => <div>{item.toString()}</div>,
      renderLoader: (state) => <div>{state.loading && 'loading...'}</div>,
      loadMore: fn,
    };
    const component = shallow(<List {...props} />);
    component.setState({
      loading: true,
    });
    expect(component).toMatchSnapshot();
  });

  it('handleLoadMore, status', () => {
    const fn = jest.fn();
    const props = {
      items: data,
      itemWidth: 60,
      indexKey: (item) => Number(item),
      renderItem: (item) => <div>{item.toString()}</div>,
      renderLoader: (state) => <div>{state.loading && 'loading...'}</div>,
      loadMore: fn,
      hasMore: true,
    };
    const component = shallow(<List {...props} />);
    component.instance().handleLoadMore();
    expect(fn).toBeCalled();

    component.setState({ loading: true });
    component.instance().handleLoadMore();
    component.instance().handleLoadMore();
    component.instance().handleLoadMore();
    expect(fn).toHaveBeenCalledTimes(1);

    component.setState({ loading: false });
    component.instance().handleLoadMore();
    expect(fn).toHaveBeenCalledTimes(2);

    component.setState({ loading: false });
    component.setProps({ hasMore: false });
    component.instance().handleLoadMore();
    expect(fn).toHaveBeenCalledTimes(2);
  });

  it('handleLoadMore, scroll down to trigger', () => {
    const fn = jest.fn();
    const props = {
      items: data,
      itemWidth: 60,
      indexKey: (item) => Number(item),
      renderItem: (item) => <div>{item.toString()}</div>,
      renderLoader: (state) => <div>{state.loading && 'loading...'}</div>,
      loadMore: fn,
      hasMore: true,
    };

    const component = shallow(<List {...props} />);
    const spy = jest.spyOn(component.instance(), 'handleLoadMore');
    window.scrollLeft = 2000;
    component.instance().update();
    expect(spy).toBeCalled();
  });

  it('update', () => {
    const fn = jest.fn();
    const props = {
      items: data,
      itemWidth: 60,
      indexKey: (item) => Number(item),
      renderItem: (item) => <div>{item.toString()}</div>,
      renderLoader: (state) => <div>{state.loading && 'loading...'}</div>,
      loadMore: fn,
      hasMore: true,
    };
    const component = shallow(<List {...props} />);
    const spy = jest.spyOn(component.instance(), 'rafUpdate');
    component.instance().handleScroll();
    expect(spy).toBeCalled();
  });

  it('mount in container', () => {
    const props = {
      items: data,
      itemWidth: 60,
      indexKey: (item) => Number(item),
      renderItem: (item) => <div>{item.toString()}</div>,
      containerStyle: {
        padding: '20px',
        borderSizing: 'border-box',
        background: 'yellow',
        height: '400px',
        overflowY: 'scroll',
      },
    };
    const component = mount(<List {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('mount document body', () => {
    const fn = jest.fn();
    const props = {
      items: data,
      itemWidth: 60,
      indexKey: (item) => Number(item),
      renderItem: (item) => <div>{item.toString()}</div>,
      renderLoader: (state) => <div>{state.loading && 'loading...'}</div>,
      loadMore: fn,
      hasMore: true,
    };

    const component = mount(<List {...props} />);
    expect(component).toMatchSnapshot();

    // scrolling
    window.scrollLeft = 2000;
    component.instance().update();
    expect(component).toMatchSnapshot();
  });
});
