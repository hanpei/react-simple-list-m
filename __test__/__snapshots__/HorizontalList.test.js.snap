// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalList mount document body 1`] = `
<HorizontalList
  containerStyle={null}
  hasMore={true}
  indexKey={[Function]}
  itemWidth={60}
  items={
    Array [
      "0",
      "1",
      "2",
      "3",
      "4",
      "5",
      "6",
      "7",
      "8",
      "9",
      "10",
      "11",
      "12",
      "13",
      "14",
      "15",
      "16",
      "17",
      "18",
      "19",
      "20",
      "21",
      "22",
      "23",
      "24",
      "25",
      "26",
      "27",
      "28",
      "29",
    ]
  }
  loadMore={[MockFunction]}
  renderItem={[Function]}
  renderLoader={[Function]}
>
  <div
    style={
      Object {
        "overflowX": "scroll",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "display": "flex",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "paddingLeft": "0px",
          "width": "1800px",
        }
      }
    />
    <div />
  </div>
</HorizontalList>
`;

exports[`HorizontalList mount document body 2`] = `
<HorizontalList
  containerStyle={null}
  hasMore={true}
  indexKey={[Function]}
  itemWidth={60}
  items={
    Array [
      "0",
      "1",
      "2",
      "3",
      "4",
      "5",
      "6",
      "7",
      "8",
      "9",
      "10",
      "11",
      "12",
      "13",
      "14",
      "15",
      "16",
      "17",
      "18",
      "19",
      "20",
      "21",
      "22",
      "23",
      "24",
      "25",
      "26",
      "27",
      "28",
      "29",
    ]
  }
  loadMore={[MockFunction]}
  renderItem={[Function]}
  renderLoader={[Function]}
>
  <div
    style={
      Object {
        "overflowX": "scroll",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "display": "flex",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "paddingLeft": "0px",
          "width": "1800px",
        }
      }
    />
    <div />
  </div>
</HorizontalList>
`;

exports[`HorizontalList mount in container 1`] = `
<HorizontalList
  containerStyle={
    Object {
      "background": "yellow",
      "borderSizing": "border-box",
      "height": "400px",
      "overflowY": "scroll",
      "padding": "20px",
    }
  }
  indexKey={[Function]}
  itemWidth={60}
  items={
    Array [
      "0",
      "1",
      "2",
      "3",
      "4",
      "5",
      "6",
      "7",
      "8",
      "9",
      "10",
      "11",
      "12",
      "13",
      "14",
      "15",
      "16",
      "17",
      "18",
      "19",
      "20",
      "21",
      "22",
      "23",
      "24",
      "25",
      "26",
      "27",
      "28",
      "29",
    ]
  }
  renderItem={[Function]}
>
  <div
    style={
      Object {
        "background": "yellow",
        "borderSizing": "border-box",
        "height": "400px",
        "overflowX": "scroll",
        "overflowY": "scroll",
        "padding": "20px",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "display": "flex",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "paddingLeft": "0px",
          "width": "1800px",
        }
      }
    />
  </div>
</HorizontalList>
`;

exports[`HorizontalList render 1`] = `
<div
  style={
    Object {
      "overflowX": "scroll",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "paddingLeft": "0px",
        "width": "1800px",
      }
    }
  >
    <div
      key="0"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        0
      </div>
    </div>
    <div
      key="1"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        1
      </div>
    </div>
    <div
      key="2"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        2
      </div>
    </div>
    <div
      key="3"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        3
      </div>
    </div>
    <div
      key="4"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        4
      </div>
    </div>
    <div
      key="5"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        5
      </div>
    </div>
    <div
      key="6"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        6
      </div>
    </div>
    <div
      key="7"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        7
      </div>
    </div>
    <div
      key="8"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        8
      </div>
    </div>
    <div
      key="9"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        9
      </div>
    </div>
    <div
      key="10"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        10
      </div>
    </div>
    <div
      key="11"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        11
      </div>
    </div>
    <div
      key="12"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        12
      </div>
    </div>
    <div
      key="13"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        13
      </div>
    </div>
    <div
      key="14"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        14
      </div>
    </div>
    <div
      key="15"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        15
      </div>
    </div>
    <div
      key="16"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        16
      </div>
    </div>
    <div
      key="17"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        17
      </div>
    </div>
    <div
      key="18"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        18
      </div>
    </div>
  </div>
</div>
`;

exports[`HorizontalList render 2`] = `
<div
  style={
    Object {
      "overflowX": "scroll",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "paddingLeft": "0px",
        "width": "2040px",
      }
    }
  >
    <div
      key="0"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        0
      </div>
    </div>
    <div
      key="1"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        1
      </div>
    </div>
    <div
      key="2"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        2
      </div>
    </div>
    <div
      key="3"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        3
      </div>
    </div>
    <div
      key="4"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        4
      </div>
    </div>
    <div
      key="5"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        5
      </div>
    </div>
    <div
      key="6"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        6
      </div>
    </div>
    <div
      key="7"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        7
      </div>
    </div>
    <div
      key="8"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        8
      </div>
    </div>
    <div
      key="9"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        9
      </div>
    </div>
    <div
      key="10"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        10
      </div>
    </div>
    <div
      key="11"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        11
      </div>
    </div>
    <div
      key="12"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        12
      </div>
    </div>
    <div
      key="13"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        13
      </div>
    </div>
    <div
      key="14"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        14
      </div>
    </div>
    <div
      key="15"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        15
      </div>
    </div>
    <div
      key="16"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        16
      </div>
    </div>
    <div
      key="17"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        17
      </div>
    </div>
    <div
      key="18"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        18
      </div>
    </div>
  </div>
</div>
`;

exports[`HorizontalList render 3`] = `""`;

exports[`HorizontalList render loading 1`] = `
<div
  style={
    Object {
      "overflowX": "scroll",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "paddingLeft": "0px",
        "width": "1800px",
      }
    }
  >
    <div
      key="0"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        0
      </div>
    </div>
    <div
      key="1"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        1
      </div>
    </div>
    <div
      key="2"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        2
      </div>
    </div>
    <div
      key="3"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        3
      </div>
    </div>
    <div
      key="4"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        4
      </div>
    </div>
    <div
      key="5"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        5
      </div>
    </div>
    <div
      key="6"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        6
      </div>
    </div>
    <div
      key="7"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        7
      </div>
    </div>
    <div
      key="8"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        8
      </div>
    </div>
    <div
      key="9"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        9
      </div>
    </div>
    <div
      key="10"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        10
      </div>
    </div>
    <div
      key="11"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        11
      </div>
    </div>
    <div
      key="12"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        12
      </div>
    </div>
    <div
      key="13"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        13
      </div>
    </div>
    <div
      key="14"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        14
      </div>
    </div>
    <div
      key="15"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        15
      </div>
    </div>
    <div
      key="16"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        16
      </div>
    </div>
    <div
      key="17"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        17
      </div>
    </div>
    <div
      key="18"
      style={
        Object {
          "boxSizing": "border-box",
          "width": "60px",
        }
      }
    >
      <div>
        18
      </div>
    </div>
  </div>
  <div>
    loading...
  </div>
</div>
`;
